feat(integrations): added authorizenet payment gateway

MyuddinKhatri opened the pull request:
added authorize payment gateway to erpnext integrations

[erpnext_documentation](https://github.com/frappe/erpnext_documentation/pull/40).

Commit:
feat(erpnext-integrations): added authorizenet payment gateway

added authorize payment gateway to erpnext integrations

Commit:
fix(requirements): added authorizenet as requirements for erpnext

added authorizenet as requirements for erpnext

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,215 @@
+	# Create a merchantAuthenticationType object with authentication details
+	merchant_auth = apicontractsv1.merchantAuthenticationType()
+	merchant_auth.name = frappe.db.get_value("Authorizenet Settings", "Authorizenet Settings", ["api_login_id"])
+	merchant_auth.transactionKey = get_decrypted_password('Authorizenet Settings', 'Authorizenet Settings',fieldname='api_transaction_key', raise_exception=False)
```
```suggestion
	# Authenticate with Authorizenet
	merchant_auth = apicontractsv1.merchantAuthenticationType()
	merchant_auth.name = frappe.db.get_single_value("Authorizenet Settings", "api_login_id")
	merchant_auth.transactionKey = get_decrypted_password('Authorizenet Settings', 'Authorizenet Settings', fieldname='api_transaction_key', raise_exception=False)
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,215 @@
+
+			# build the array of line items
+			line_items = apicontractsv1.ArrayOfLineItem()
+			line_items.lineItem.append(item[i])
```
```suggestion
	# build the array of line items
	line_items = apicontractsv1.ArrayOfLineItem()

	for item in sales_order.get("items"):
		# setup individual line items
		line_item = apicontractsv1.lineItemType()
		line_item.itemId = item.item_code
		line_item.name = item.item_name[:30]
		line_item.description = item.item_name
		line_item.quantity = item.qty
		line_item.unitPrice = item.amount
		
		line_items.lineItem.append(line_item)
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,215 @@
+
+	custom_redirect_to = None
+
+	if status != "Failed":
```
This will fail if `response` does end up being `None`. Set a default value before the response is processed.

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,215 @@
+				status = "Failed"
+
+			else:
+				status = "Failed"
```
Use the `Integration Request` to handle all the logging, see how Stripe is doing it

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,215 @@
+
+	if status != "Failed":
+		try:
+			custom_redirect_to = frappe.get_doc(data.reference_doctype, data.reference_docname).run_method("on_payment_authorized",
```
You're not using this variable anywhere

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,215 @@
+			raise ex
+
+	response = to_dict(response)
+	transId = response.get("transactionResponse").get("transId")
```
Can we not just do `transId = response.transactionResponse.transId`?

Authorizenet's documentation seems to use that

Commit:
fix(erpnext-integrations): fixed bugs in authorizenet_settings.py

fixed bugs in authoriznet_settings.py

Commit:
feat(erpnext-integrations): added feature to create integration request

added feature to create integration request for payment entry and fixed html css

Commit:
fix(erpnext-integrations): added maxlength to card expiry

added maxlength to card expiry

Commit:
feat(erpnext-integrations): added error response to integration request

added error response to integration request

Commit:
format: checkout page

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+	payment = apicontractsv1.paymentType()
+	payment.creditCard = credit_card
+
+	pr = frappe.get_doc("Payment Request", data.reference_docname)
```
```suggestion
	pr = frappe.get_doc(data.reference_doctype, data.reference_docname)
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+			item[i] = apicontractsv1.lineItemType()
+			item[i].itemId = item.item_code
+			item[i].name = item.item_name[:30]
+			item[i].description = item.item_name
```
Confirm if there's a limit on this key, cause descriptions can get long

```suggestion
			item[i].description = item.description
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+			item[i].name = item.item_name[:30]
+			item[i].description = item.item_name
+			item[i].quantity = item.qty
+			item[i].unitPrice = item.amount
```
Amount is total, this will be per-unit price

```suggestion
			item[i].unitPrice = item.rate
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+
+	if status != "Failed":
+		try:
+			frappe.get_doc(data.reference_doctype, data.reference_docname).run_method("on_payment_authorized", status)
```
```suggestion
			pr.run_method("on_payment_authorized", status)
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+	elif status == "Failed":
+		error_text = response_dict.get("transactionResponse").get("errors").get("error").get("errorText")
+		description = error_text
+		doc.update_status(data, status)
```
Pull this out of the conditions, since this needs to happen anyway

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+
+	response = createtransactioncontroller.getresponse()
+
+	doc = frappe.get_doc("Integration Request", integration_request.name)
```
You don't need this, you already have the `integration_request` object

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+		error_text = response_dict.get("transactionResponse").get("errors").get("error").get("errorText")
+		description = error_text
+		doc.update_status(data, status)
+		frappe.db.set_value("Integration Request", doc.name, "error", error_text)
```
```suggestion
		doc.error = error_text
		doc.save(ignore_permissions=True)
```

Alchez commented on erpnext/erpnext_integrations/doctype/authorizenet_settings/authorizenet_settings.py:
```diff
@@ -0,0 +1,216 @@
+
+	doc = frappe.get_doc("Integration Request", integration_request.name)
+
+	if response is not None:
+		# Check to see if the API request was successfully received and acted upon
+		if response.messages.resultCode == "Ok":
+			# Since the API request was successful, look for a transaction response
+			# and parse it to display the results of authorizing the card
+			if hasattr(response.transactionResponse, 'messages') is True:
+				status = "Completed"
+			else:
+				status = "Failed"
+				if hasattr(response.transactionResponse, 'errors') is True:
+					status = "Failed"
+
+		# Or, print errors if the API request wasn't successful
+		else:
+			status = "Failed"
+			if hasattr(response, 'transactionResponse') is True and hasattr(
+					response.transactionResponse, 'errors') is True:
+				status = "Failed"
+
+			else:
+				status = "Failed"
```
```suggestion
	status = "Failed"
	if response is not None:
		if response.messages.resultCode == "Ok" and hasattr(response.transactionResponse, 'messages') is True:
			status = "Completed"
```

Commit:
fix(erpnext-integrations): added cardnumber validation

added cardnumber validation

Commit:
fixed minor changes

Alchez commented:
Made upstream to [develop](https://github.com/frappe/erpnext/pull/21437).
